pkg/kgo/txn.go

```diff
@@ -1060,7 +1061,13 @@ func (cl *Client) commitTransactionOffsets(
 		onDone(kmsg.NewPtrTxnOffsetCommitRequest(), kmsg.NewPtrTxnOffsetCommitResponse(), errNotGroup)
 		return nil
 	}
-	if len(uncommitted) == 0 {
+
+	req, err := g.prepareTxnOffsetCommit(ctx, uncommitted)
```

Comment:
This needs to be in g.mu because building the request uses the g.generation and g.memberID fields, which are updated after join group (within a lock). Need to go, will look at this in a bit to see where it can go exactly unless you eyeball this very closely.